File: src/turingarena/driver/languages/python/sandbox.py
import errno
import sys

try:
    import seccomplite
except ImportError:
    seccomplite = None
    print("WARNING: no code sandboxing! Install seccomplite")


def init_sandbox():
    filter = seccomplite.Filter(seccomplite.ERRNO(errno.EACCES))
    filter.add_rule(seccomplite.ALLOW, "read", seccomplite.Arg(0, seccomplite.EQ, 0))
    filter.add_rule(seccomplite.ALLOW, "write", seccomplite.Arg(0, seccomplite.EQ, 1))
    filter.add_rule(seccomplite.ALLOW, "write", seccomplite.Arg(0, seccomplite.EQ, 2))
    filter.add_rule(seccomplite.ALLOW, "exit")
    filter.add_rule(seccomplite.ALLOW, "exit_group")
    filter.add_rule(seccomplite.ALLOW, "rt_sigreturn")
    filter.add_rule(seccomplite.ALLOW, "mmap")
    filter.add_rule(seccomplite.ALLOW, "munmap")
    filter.add_rule(seccomplite.ALLOW, "mremap")
    filter.add_rule(seccomplite.ALLOW, "brk")
    filter.add_rule(seccomplite.ALLOW, "lseek")
    filter.add_rule(seccomplite.ALLOW, "ioctl")
    filter.load()


def main():
    source_path, skeleton_path = sys.argv[1:]

    with open(source_path) as source_file:
        source_string = source_file.read()
    with open(skeleton_path) as skeleton_file:
        skeleton_string = skeleton_file.read()

    if seccomplite is not None:
        init_sandbox()

    class Wrapper: pass

    skeleton = Wrapper()
    source = Wrapper()

    # run skeleton and source
    exec(source_string, source.__dict__)
    exec(skeleton_string, skeleton.__dict__)

    skeleton.main(source)


if __name__ == "__main__":
    main()

File: src/turingarena/driver/languages/ruby/sandbox.rb
require 'fiddle/import'

module Seccomp
  extend Fiddle::Importer
  dlload "libseccomp.so.2"
  extern "void* seccomp_init(int)"
  extern "int seccomp_rule_add(void*, int, int, int)"
  extern "int seccomp_load(void*)"
  extern "int seccomp_syscall_resolve_name(void*)"

  SCMP_ACT_KILL     = 0x00000000
  SCMP_ACT_TRAP     = 0x00030000
  SCMP_ACT_ERRNO_0  = 0x00050000
  SCMP_ACT_ALLOW    = 0x7fff0000
end

def init_sandbox()
    ctx = Seccomp.seccomp_init(Seccomp::SCMP_ACT_TRAP)
    ret = 0
    syscalls = [
        "read", "write", "readv", "writev",  # base I/O
        "lseek", "ioctl", "fstat",
        "exit", "exit_group", "rt_sigreturn",
        "mmap", "munmap", "mremap", "brk",
        "execve",
        "arch_prctl", "uname", "set_tid_address",
        "time", "futex", "close", "getpid"
    ]
    syscalls.each do |s|
        ret |= Seccomp::seccomp_rule_add(ctx, Seccomp::SCMP_ACT_ALLOW, Seccomp::seccomp_syscall_resolve_name(s), 0)
    end

    ret |= Seccomp::seccomp_load(ctx)
    fail "Something went wrong." unless ret == 0
end

def main()
    load ARGV[0]
    init_sandbox

    # From here every system call that is not allowed will result in an Exception

    main
end

if __FILE__ == $0
    main
end
